packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -97,28 +121,31 @@ export class DefaultCompletionProvider implements CompletionProvider {
         await Promise.all(
             stream(beforeFeatures)
                 .distinct(distinctionFunction)
-                .map(e => this.completionFor(node.element, e, acceptor))
+                .map(e => this.completionFor(context, e, acceptor))
         );
 
         if (reparse) {
-            const missingPart = textDocument.getText({
-                start: textDocument.positionAt(parserStart),
-                end: params.position
-            }).toLowerCase();
-            // Remove items from `beforeFeatures` which don't fit the current text
-            items = items.filter(e => e.label.toLowerCase().startsWith(missingPart));
-
             await Promise.all(
                 stream(afterFeatures)
                     .exclude(beforeFeatures, distinctionFunction)
                     .distinct(distinctionFunction)
-                    .map(e => this.completionFor(node.element, e, acceptor))
+                    .map(e => this.completionFor(context, e, acceptor))
             );
         }
 
         return CompletionList.create(items, true);
     }
 
+    /**
+     * Determines whether the completion parser will reparse the input at the point of completion.
+     * By default, this returns `false`, indicating that the completion will only look for completion results starting from the token at the cursor position.
+     * Override this and return `true` to indicate that the completion should parse the input a second time.
+     * This might add some missing completions at the cost at parsing the input twice.
```

Comment:
Please add a test so we know that we don't accidentally break this functionality (it's disabled by default), or create an issue to add such a test.

Reply:
Done see https://github.com/langium/langium/issues/831